frmts/heif/heifdataset.cpp

```diff
@@ -637,14 +655,88 @@ GDALHEIFRasterBand::GDALHEIFRasterBand(GDALHEIFDataset *poDSIn, int nBandIn)
                                         "IMAGE_STRUCTURE");
     }
 #endif
+
+#ifdef LIBHEIF_SUPPORTS_TILES
+    nBlockXSize = poDSIn->m_tiling.tile_width;
+    nBlockYSize = poDSIn->m_tiling.tile_height;
+#else
     nBlockXSize = poDS->GetRasterXSize();
     nBlockYSize = 1;
+#endif
 }
 
 /************************************************************************/
 /*                            IReadBlock()                              */
 /************************************************************************/
+#ifdef LIBHEIF_SUPPORTS_TILES
+CPLErr GDALHEIFRasterBand::IReadBlock(int nBlockXOff, int nBlockYOff,
+                                      void *pImage)
+{
+    GDALHEIFDataset *poGDS = static_cast<GDALHEIFDataset *>(poDS);
+    if (poGDS->m_bFailureDecoding)
+        return CE_Failure;
+    const int nBands = poGDS->GetRasterCount();
+    heif_image *hImage = nullptr;
+    struct heif_decoding_options *decode_options =
+        heif_decoding_options_alloc();
+
+    auto err = heif_image_handle_decode_image_tile(
+        poGDS->m_hImageHandle, &hImage, heif_colorspace_RGB,
+        nBands == 3
+            ? (eDataType == GDT_UInt16 ?
+#if CPL_IS_LSB
+                                       heif_chroma_interleaved_RRGGBB_LE
+#else
+                                       heif_chroma_interleaved_RRGGBB_BE
+#endif
+                                       : heif_chroma_interleaved_RGB)
+            : (eDataType == GDT_UInt16 ?
+#if CPL_IS_LSB
+                                       heif_chroma_interleaved_RRGGBBAA_LE
+#else
+                                       heif_chroma_interleaved_RRGGBBAA_BE
+#endif
+                                       : heif_chroma_interleaved_RGBA),
+        decode_options, nBlockXOff, nBlockYOff);
 
+    if (err.code != heif_error_Ok)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "%s",
+                 err.message ? err.message : "Cannot decode image");
+        poGDS->m_bFailureDecoding = true;
+        return CE_Failure;
+    }
+    heif_decoding_options_free(decode_options);
+    int nStride = 0;
+    const uint8_t *pSrcData = heif_image_get_plane_readonly(
+        hImage, heif_channel_interleaved, &nStride);
+    if (eDataType == GDT_Byte)
+    {
+        for (int y = 0; y < nBlockYSize; y++)
+        {
+            for (int x = 0; x < nBlockXSize; x++)
+            {
+                (static_cast<GByte *>(pImage))[y * nBlockXSize + x] =
```

Comment:
is the a risk that nBlockXSize * nBlockYSize might overflow INT_MAX ? If so, then y * nBlockXSize + x here and in the else clause below should be rewritten as ``static_cast<size_t>(y) * nBlockXSize + x``

Reply:
Should be fixed.

Out of interest, would defining `x` and `y` to be `size_t` instead of `int` be an acceptable alternative?

Reply:
you could do that as well. up to you. You'll just get a compiler warning when comparing x/y to nBlockX/YSize about mismatched signedness, requiring a cast of nBlockX/YSize to size_t